src/Xamarin.Android.Build.Tasks/Tasks/GeneratePackageManagerJava.cs

```diff
@@ -103,10 +131,133 @@ public override bool Execute ()
+				environment.WriteLine ("\t\t// Source File: {0}", env.ItemSpec);
+				foreach (string line in File.ReadLines (env.ItemSpec)) {
+					var lineToWrite = line;
+					if (lineToWrite.StartsWith ("MONO_LOG_LEVEL=", StringComparison.Ordinal))
+						haveLogLevel = true;
+					if (lineToWrite.StartsWith ("MONO_GC_PARAMS=", StringComparison.Ordinal))
+						haveMonoGCParams = true;
+					if (lineToWrite.StartsWith ("XAMARIN_BUILD_ID=", StringComparison.Ordinal))
+						havebuildId = true;
+					if (lineToWrite.StartsWith ("MONO_DEBUG=", StringComparison.Ordinal)) {
+						haveMonoDebug = true;
+						if (sequencePointsMode != SequencePointsMode.None && !lineToWrite.Contains ("gen-compact-seq-points"))
+							lineToWrite = line  + ",gen-compact-seq-points";
+					}
+					if (lineToWrite.StartsWith ("XA_HTTP_CLIENT_HANDLER_TYPE=", StringComparison.Ordinal))
+						haveHttpMessageHandler = true;
+					if (lineToWrite.StartsWith ("XA_TLS_PROVIDER=", StringComparison.Ordinal))
+						haveTlsProvider = true;
+					WriteEnvironmentLine (lineToWrite);
+				}
+			}
+
+			if (_Debug && !haveLogLevel) {
+				WriteEnvironment (defaultLogLevel[0], defaultLogLevel[1]);
+			}
+
+			if (sequencePointsMode != SequencePointsMode.None && !haveMonoDebug) {
+				WriteEnvironment (defaultMonoDebug[0], defaultMonoDebug[1]);
+			}
+
+			if (!havebuildId)
+				WriteEnvironment ("XAMARIN_BUILD_ID", buildId.ToString ());
+
+			if (!haveHttpMessageHandler) {
+				if (HttpClientHandlerType == null)
+					WriteEnvironment (defaultHttpMessageHandler[0], defaultHttpMessageHandler[1]);
+				else
+					WriteEnvironment ("XA_HTTP_CLIENT_HANDLER_TYPE", HttpClientHandlerType.Trim ());
+			}
+
+			if (!haveTlsProvider) {
+				if (TlsProvider == null)
+					WriteEnvironment (defaultTlsProvider[0], defaultTlsProvider[1]);
+				else
+					WriteEnvironment ("XA_TLS_PROVIDER", TlsProvider.Trim ());
+			}
+
+			if (!haveMonoGCParams) {
+				if (EnableSGenConcurrent)
+					WriteEnvironment ("MONO_GC_PARAMS", "major=marksweep-conc");
+				else
+					WriteEnvironment ("MONO_GC_PARAMS", "major=marksweep");
+			}
+
+			string environmentTemplate;
+			using (var sr = new StreamReader (typeof (BuildApk).Assembly.GetManifestResourceStream (EnvironmentFileName))) {
+				environmentTemplate = sr.ReadToEnd ();
+			}
+
+			string temp = Path.GetTempFileName ();
+			using (var fs = File.CreateText (temp)) {
+				fs.Write (environmentTemplate.Replace ("//@ENVVARS@", environment.ToString ()));
+			}
+
+			string dest = Path.GetFullPath (Path.Combine (EnvironmentOutputDirectory, EnvironmentFileName));
+			MonoAndroidHelper.CopyIfChanged (temp, dest);
+			try { File.Delete (temp); } catch (Exception) { }
+
+			void WriteEnvironment (string name, string value)
+			{
+				environment.WriteLine ($"\t\t\"{ValidJavaString (name)}\", \"{ValidJavaString (value)}\",");
+			}
+
+			void WriteEnvironmentLine (string line)
+			{
+				if (String.IsNullOrEmpty (line))
+					return;
+
+				string[] nv = line.Split ('=');
```

Comment:
This should use `line.Split(new char[]{'='}, 2)`; see also:

```
$ git grep Split.*2 | grep .cs:
build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/RunInstrumentationTests.cs:				string[] kvp = a.Split (new [] { '=' }, 2);
build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ReplaceFileContents.cs:				var kvp = e.Split (Separator, 2, StringSplitOptions.RemoveEmptyEntries);
src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/MarkJavaObjects.cs:				var values = methodAndType.Split (new char [] { ':' }, 2);
src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/PreserveHttpAndroidClientHandler.cs:			var parts = typeName.Split (new char [] { ',' }, 2);
src/Xamarin.Android.Build.Tasks/Utilities/AndroidAddOnManifest.cs:					string[] keyValue = line.Split (new char[]{'='}, 2);
src/Xamarin.Android.Build.Tasks/Utilities/AndroidAddOnManifest.cs:							string filename = keyValue [1].Split (new []{';'}, 2) [0];
src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocument.cs:				foreach (var entry in Placeholders.Select (e => e.Split (new char [] {'='}, 2, StringSplitOptions.None))) {
src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:				var items = s.Split (new char[] { ';' }, count: 2);
src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:				var items = s.Split (new char [] { ';' }, count: 2);
src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:				foreach (var arr in resourceCaseMap.Split (';').Select (l => l.Split ('|')).Where (a => a.Length == 2))
```

This avoids the need for your `string.Join ("=", ...)` below.

Reply:
I never liked unnecessary allocations like those shown in this pattern.